File: ios/StreamSessionManager.swift
import Foundation
import UIKit
import MWDATCore
import MWDATCamera

/// Callback type for frame updates
public typealias FrameCallback = (UIImage) -> Void

/// Manages camera streaming as a capability attached to a DeviceSession.
@MainActor
public final class StreamSessionManager {
    public static let shared = StreamSessionManager()

    private let logger = EMWDATLogger.shared

    // MARK: - State

    /// Active stream sessions keyed by sessionId
    private var streams: [String: StreamSession] = [:]
    private var stateTokens: [String: AnyListenerToken] = [:]
    private var frameTokens: [String: AnyListenerToken] = [:]
    private var errorTokens: [String: AnyListenerToken] = [:]
    private var photoTokens: [String: AnyListenerToken] = [:]
    private var hevcDecoders: [String: HEVCDecoder] = [:]

    // MARK: - Callbacks

    private var eventEmitter: EventEmitter?
    private var frameCallback: FrameCallback?
    private var frameCallbackOwner: UUID?

    private init() {}

    // MARK: - Configuration

    /// Set the event emitter for sending events to JavaScript
    public func setEventEmitter(_ emitter: @escaping EventEmitter) {
        self.eventEmitter = emitter
    }

    /// Set the frame callback for native view rendering
    public func setFrameCallback(_ callback: @escaping FrameCallback, owner: UUID) {
        self.frameCallback = callback
        self.frameCallbackOwner = owner
    }

    /// Remove the frame callback only if the caller is the current owner
    public func removeFrameCallback(owner: UUID) {
        guard frameCallbackOwner == owner else { return }
        self.frameCallback = nil
        self.frameCallbackOwner = nil
    }

    // MARK: - Stream Capability Control

    /// Add a camera stream capability to a device session and start streaming.
    public func addStreamToSession(sessionId: String, config: StreamSessionConfig) async throws {
        guard let session = WearablesManager.shared.getSession(sessionId: sessionId) else {
            throw StreamSessionManagerError.sessionNotFound(sessionId)
        }

        logger.info("StreamSession", "Adding stream to session", context: [
            "sessionId": sessionId,
            "resolution": String(describing: config.resolution),
            "frameRate": config.frameRate,
            "codec": String(describing: config.videoCodec)
        ])

        // Create HEVC decoder if needed
        if config.videoCodec == .hvc1 {
            hevcDecoders[sessionId] = HEVCDecoder()
            logger.info("StreamSession", "HEVC decoder created for session", context: ["sessionId": sessionId])
        }

        // Add stream capability to the session
        guard let streamSession = try session.addStream(config: config) else {
            throw StreamSessionManagerError.streamNotFound(sessionId)
        }
        streams[sessionId] = streamSession

        // Subscribe to state changes
        stateTokens[sessionId] = streamSession.statePublisher.listen { [weak self] state in
            Task { @MainActor in
                self?.handleStateChange(sessionId: sessionId, state: state)
            }
        }

        // Subscribe to video frames
        frameTokens[sessionId] = streamSession.videoFramePublisher.listen { [weak self] frame in
            Task { @MainActor in
                self?.handleVideoFrame(sessionId: sessionId, frame: frame)
            }
        }

        // Subscribe to errors
        errorTokens[sessionId] = streamSession.errorPublisher.listen { [weak self] error in
            Task { @MainActor in
                self?.handleError(sessionId: sessionId, error: error)
            }
        }

        // Subscribe to photos
        photoTokens[sessionId] = streamSession.photoDataPublisher.listen { [weak self] photoData in
            Task { @MainActor in
                self?.handlePhotoCapture(sessionId: sessionId, photoData: photoData)
            }
        }

        // Start the stream (required in SDK v0.6)
        await streamSession.start()

        // Emit initial capability state
        emitEvent("onCapabilityStateChange", [
            "sessionId": sessionId,
            "state": "active"
        ])

        logger.info("StreamSession", "Stream added and started", context: ["sessionId": sessionId])
    }

    /// Remove the camera stream capability from a session.
    public func removeStreamFromSession(sessionId: String) async {
        // Stop the stream before destroying
        await streams[sessionId]?.stop()

        destroyStream(sessionId: sessionId)

        emitEvent("onCapabilityStateChange", [
            "sessionId": sessionId,
            "state": "stopped"
        ])

        logger.info("StreamSession", "Stream removed from session", context: ["sessionId": sessionId])
    }

    /// Capture a photo from the active stream on any session.
    public func capturePhoto(format: PhotoCaptureFormat) -> Bool {
        // Find the first streaming session
        guard let (_, streamSession) = streams.first(where: { _ in true }) else {
            logger.warn("StreamSession", "Cannot capture photo - no active stream")
            return false
        }

        logger.info("StreamSession", "Capturing photo", context: ["format": String(describing: format)])
        return streamSession.capturePhoto(format: format)
    }

    // MARK: - Event Handlers

    private func handleStateChange(sessionId: String, state: StreamSessionState) {
        logger.info("StreamSession", "State changed", context: [
            "sessionId": sessionId,
            "state": String(describing: state)
        ])

        emitEvent("onStreamStateChange", [
            "state": mapStreamState(state)
        ])
    }

    private func handleVideoFrame(sessionId: String, frame: VideoFrame) {
        let decoder = hevcDecoders[sessionId]

        // Try SDK's built-in conversion first (works for raw codec).
        // For HEVC, makeUIImage() returns nil — fall back to hardware decoder.
        let image: UIImage
        if let direct = frame.makeUIImage() {
            image = direct
        } else if let decoded = decoder?.decode(frame.sampleBuffer) {
            image = decoded
        } else {
            return
        }

        // Forward to native view
        frameCallback?(image)

        // Emit metadata to JS
        emitEvent("onVideoFrame", [
            "timestamp": Int(Date().timeIntervalSince1970 * 1000),
            "width": Int(image.size.width),
            "height": Int(image.size.height)
        ])
    }

    private func handleError(sessionId: String, error: StreamSessionError) {
        logger.error("StreamSession", "Stream error", context: [
            "sessionId": sessionId,
            "error": String(describing: error)
        ])

        emitEvent("onStreamError", mapStreamErrorToDict(error))
    }

    private func handlePhotoCapture(sessionId: String, photoData: PhotoData) {
        logger.info("StreamSession", "Photo captured", context: [
            "sessionId": sessionId,
            "format": String(describing: photoData.format)
        ])

        let tempDir = FileManager.default.temporaryDirectory
        let ext = photoData.format == .jpeg ? "jpg" : "heic"
        let filename = "emwdat_photo_\(Int(Date().timeIntervalSince1970 * 1000)).\(ext)"
        let filePath = tempDir.appendingPathComponent(filename)

        do {
            try photoData.data.write(to: filePath)
            logger.info("StreamSession", "Photo saved", context: ["path": filePath.path])

            var payload: [String: Any] = [
                "filePath": filePath.path,
                "format": mapPhotoFormat(photoData.format),
                "timestamp": Int(Date().timeIntervalSince1970 * 1000)
            ]

            if let image = UIImage(data: photoData.data) {
                payload["width"] = Int(image.size.width * image.scale)
                payload["height"] = Int(image.size.height * image.scale)
            }

            emitEvent("onPhotoCaptured", payload)
        } catch {
            logger.error("StreamSession", "Failed to save photo", error: error)
        }
    }

    // MARK: - Cleanup

    /// Destroy a specific stream (listeners + decoder).
    private func destroyStream(sessionId: String) {
        stateTokens[sessionId] = nil
        frameTokens[sessionId] = nil
        errorTokens[sessionId] = nil
        photoTokens[sessionId] = nil
        streams[sessionId] = nil
        hevcDecoders[sessionId]?.invalidate()
        hevcDecoders[sessionId] = nil
        logger.debug("StreamSession", "Stream destroyed", context: ["sessionId": sessionId])
    }

    /// Full teardown for module lifecycle (OnDestroy).
    public func destroy() {
        for sessionId in streams.keys {
            destroyStream(sessionId: sessionId)
        }
    }

    // MARK: - Event Emission

    private func emitEvent(_ name: String, _ body: [String: Any]) {
        eventEmitter?(name, body)
    }

    // MARK: - Mapping Helpers

    private func mapStreamState(_ state: StreamSessionState) -> String {
        switch state {
        case .stopping: return "stopping"
        case .stopped: return "stopped"
        case .waitingForDevice: return "waitingForDevice"
        case .starting: return "starting"
        case .streaming: return "streaming"
        case .paused: return "paused"
        @unknown default: return "stopped"
        }
    }

    /// Maps StreamSessionError to a dictionary matching the TS discriminated union
    private func mapStreamErrorToDict(_ error: StreamSessionError) -> [String: Any] {
        switch error {
        case .deviceNotFound(let deviceId):
            return ["type": "deviceNotFound", "deviceId": deviceId]
        case .deviceNotConnected(let deviceId):
            return ["type": "deviceNotConnected", "deviceId": deviceId]
        case .timeout:
            return ["type": "timeout"]
        case .permissionDenied:
            return ["type": "permissionDenied"]
        case .internalError:
            return ["type": "internalError"]
        case .videoStreamingError:
            return ["type": "videoStreamingError"]
        case .hingesClosed:
            return ["type": "hingesClosed"]
        case .thermalCritical:
            return ["type": "thermalCritical"]
        @unknown default:
            return ["type": "internalError"]
        }
    }

    private func mapPhotoFormat(_ format: PhotoCaptureFormat) -> String {
        switch format {
        case .jpeg: return "jpeg"
        case .heic: return "heic"
        @unknown default: return "jpeg"
        }
    }
}

// MARK: - Configuration Parsing

extension StreamSessionManager {
    /// Parse configuration from JavaScript object
    nonisolated public static func parseConfig(from dict: [String: Any]) -> StreamSessionConfig {
        let videoCodec: VideoCodec
        let compressVideo = dict["compressVideo"] as? Bool ?? false
        if compressVideo || (dict["videoCodec"] as? String) == "hvc1" {
            videoCodec = .hvc1
        } else {
            videoCodec = .raw
        }

        let resolution: StreamingResolution
        if let resStr = dict["resolution"] as? String {
            switch resStr {
            case "high": resolution = .high
            case "medium": resolution = .medium
            default: resolution = .low
            }
        } else {
            resolution = .low
        }

        let frameRate = dict["frameRate"] as? Int ?? 15
        let skipAppLaunch = dict["skipAppLaunch"] as? Bool ?? false

        return StreamSessionConfig(
            videoCodec: videoCodec,
            resolution: resolution,
            frameRate: UInt(frameRate),
            skipAppLaunch: skipAppLaunch
        )
    }
}

// MARK: - Errors

public enum StreamSessionManagerError: LocalizedError {
    case sessionNotFound(String)
    case streamNotFound(String)
    case notConfigured

    public var errorDescription: String? {
        switch self {
        case .sessionNotFound(let id):
            return "Session not found: \(id)"
        case .streamNotFound(let id):
            return "No active stream for session: \(id)"
        case .notConfigured:
            return "Wearables SDK has not been configured"
        }
    }
}

File: android/src/main/java/expo/modules/emwdat/StreamSessionManager.kt
package expo.modules.emwdat

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import androidx.exifinterface.media.ExifInterface
import com.meta.wearable.dat.camera.Stream
import com.meta.wearable.dat.camera.types.CaptureError
import com.meta.wearable.dat.camera.types.PhotoData
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamSessionState
import com.meta.wearable.dat.camera.types.VideoFrame
import com.meta.wearable.dat.camera.types.VideoQuality
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File

typealias FrameCallback = (Bitmap) -> Unit

object StreamSessionManager {
    private val logger = EMWDATLogger

    // Active streams keyed by sessionId
    private val streams: MutableMap<String, Stream> = mutableMapOf()
    private val videoJobs: MutableMap<String, Job> = mutableMapOf()
    private val stateJobs: MutableMap<String, Job> = mutableMapOf()
    private val errorJobs: MutableMap<String, Job> = mutableMapOf()

    private var scope: CoroutineScope? = null

    // Callbacks
    private var eventEmitter: EventEmitter? = null
    private var frameCallback: FrameCallback? = null
    private var frameCallbackOwner: Any? = null

    fun setEventEmitter(emitter: EventEmitter) {
        this.eventEmitter = emitter
    }

    fun setScope(scope: CoroutineScope) {
        this.scope = scope
    }

    fun setFrameCallback(callback: FrameCallback, owner: Any) {
        this.frameCallback = callback
        this.frameCallbackOwner = owner
    }

    fun removeFrameCallback(owner: Any) {
        if (frameCallbackOwner !== owner) return
        this.frameCallback = null
        this.frameCallbackOwner = null
    }

    // MARK: - Stream Capability Control

    fun addStreamToSession(sessionId: String, config: Map<String, Any>) {
        val session = WearablesManager.getSession(sessionId)
            ?: throw IllegalArgumentException("Session not found: $sessionId")

        val videoQuality = when (config["resolution"] as? String) {
            "high" -> VideoQuality.HIGH
            "medium" -> VideoQuality.MEDIUM
            else -> VideoQuality.LOW
        }
        val frameRate = (config["frameRate"] as? Number)?.toInt() ?: 15
        val compressVideo = config["compressVideo"] as? Boolean ?: false

        val streamConfig = StreamConfiguration(videoQuality, frameRate, compressVideo)

        logger.info("StreamSession", "Adding stream to session", mapOf(
            "sessionId" to sessionId,
            "quality" to videoQuality.toString(),
            "frameRate" to frameRate,
            "compressVideo" to compressVideo
        ))

        val stream = session.addStream(streamConfig)
        streams[sessionId] = stream

        val currentScope = scope ?: throw IllegalStateException("Module scope not available")

        // Collect video frames
        videoJobs[sessionId] = currentScope.launch {
            stream.videoStream.collect { frame ->
                handleVideoFrame(sessionId, frame)
            }
        }

        // Collect state changes
        stateJobs[sessionId] = currentScope.launch {
            stream.state.collect { state ->
                handleStateChange(sessionId, state)
            }
        }

        // Collect errors
        errorJobs[sessionId] = currentScope.launch {
            stream.errorStream.collect { error ->
                logger.error("StreamSession", "Stream error", mapOf(
                    "sessionId" to sessionId,
                    "error" to error.toString()
                ))
                emitEvent("onStreamError", mapOf("type" to "internalError"))
            }
        }

        // Emit capability state
        emitEvent("onCapabilityStateChange", mapOf(
            "sessionId" to sessionId,
            "state" to "active"
        ))

        logger.info("StreamSession", "Stream added to session", mapOf("sessionId" to sessionId))
    }

    fun removeStreamFromSession(sessionId: String) {
        val session = WearablesManager.getSession(sessionId)
        session?.removeStream()
        destroyStream(sessionId)

        emitEvent("onCapabilityStateChange", mapOf(
            "sessionId" to sessionId,
            "state" to "stopped"
        ))

        logger.info("StreamSession", "Stream removed from session", mapOf("sessionId" to sessionId))
    }

    suspend fun capturePhoto(context: Context, format: String) {
        // Find the first active stream
        val stream = streams.values.firstOrNull()
            ?: throw Exception("No active stream session")

        logger.info("StreamSession", "Capturing photo", mapOf("requestedFormat" to format))
        val result = stream.capturePhoto()

        result.fold(
            onSuccess = { photoData ->
                handlePhotoCapture(context, photoData, format)
            },
            onFailure = { error ->
                val msg = when (error) {
                    is CaptureError.DeviceDisconnected -> "Device disconnected"
                    is CaptureError.NotStreaming -> "Not streaming"
                    is CaptureError.CaptureInProgress -> "Capture already in progress"
                    is CaptureError.CaptureFailed -> "Capture failed"
                }
                logger.error("StreamSession", "Photo capture failed", mapOf("error" to msg))
                throw Exception("Photo capture failed: $msg")
            }
        )
    }

    // MARK: - Frame Handling

    private fun handleVideoFrame(sessionId: String, videoFrame: VideoFrame) {
        // If frame is compressed HEVC, emit metadata only (can't decode to bitmap)
        if (videoFrame.isCompressed) {
            emitEvent("onVideoFrame", mapOf(
                "timestamp" to System.currentTimeMillis(),
                "width" to videoFrame.width,
                "height" to videoFrame.height,
                "isCompressed" to true
            ))
            return
        }

        val buffer = videoFrame.buffer
        val dataSize = buffer.remaining()
        val byteArray = ByteArray(dataSize)

        val originalPosition = buffer.position()
        buffer.get(byteArray)
        buffer.position(originalPosition)

        // I420 -> NV21 -> JPEG -> Bitmap
        val nv21 = convertI420toNV21(byteArray, videoFrame.width, videoFrame.height)
        val image = YuvImage(nv21, ImageFormat.NV21, videoFrame.width, videoFrame.height, null)
        val out = ByteArrayOutputStream()
        image.compressToJpeg(Rect(0, 0, videoFrame.width, videoFrame.height), 50, out)
        val jpegBytes = out.toByteArray()
        val bitmap = BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size) ?: return

        // Forward to native view
        frameCallback?.invoke(bitmap)

        // Emit metadata to JS
        emitEvent("onVideoFrame", mapOf(
            "timestamp" to System.currentTimeMillis(),
            "width" to videoFrame.width,
            "height" to videoFrame.height,
            "isCompressed" to false
        ))
    }

    private fun convertI420toNV21(input: ByteArray, width: Int, height: Int): ByteArray {
        val output = ByteArray(input.size)
        val size = width * height
        val quarter = size / 4

        // Copy Y plane directly
        input.copyInto(output, 0, 0, size)

        // Interleave V and U planes into NV21 format
        for (n in 0 until quarter) {
            output[size + n * 2] = input[size + quarter + n]     // V
            output[size + n * 2 + 1] = input[size + n]           // U
        }
        return output
    }

    // MARK: - State Handling

    private fun handleStateChange(sessionId: String, state: StreamSessionState) {
        val mapped = mapStreamState(state)
        logger.info("StreamSession", "State changed", mapOf(
            "sessionId" to sessionId,
            "state" to mapped
        ))

        emitEvent("onStreamStateChange", mapOf("state" to mapped))
    }

    // MARK: - Photo Handling

    private fun handlePhotoCapture(context: Context, photoData: PhotoData, requestedFormat: String) {
        val timestamp = System.currentTimeMillis()
        val tempDir = context.cacheDir

        when (photoData) {
            is PhotoData.Bitmap -> {
                val filename = "emwdat_photo_${timestamp}.jpg"
                val file = File(tempDir, filename)
                file.outputStream().use { out ->
                    photoData.bitmap.compress(Bitmap.CompressFormat.JPEG, 95, out)
                }
                logger.info("StreamSession", "Photo saved (Bitmap→JPEG)", mapOf("path" to file.absolutePath))

                emitEvent("onPhotoCaptured", mapOf(
                    "filePath" to file.absolutePath,
                    "format" to "jpeg",
                    "timestamp" to timestamp,
                    "width" to photoData.bitmap.width,
                    "height" to photoData.bitmap.height
                ))
            }
            is PhotoData.HEIC -> {
                val buffer = photoData.data
                val bytes = ByteArray(buffer.remaining())
                val originalPos = buffer.position()
                buffer.get(bytes)
                buffer.position(originalPos)

                // If JPEG requested, decode HEIC to bitmap and re-encode as JPEG
                if (requestedFormat == "jpeg") {
                    val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                    if (bitmap != null) {
                        val filename = "emwdat_photo_${timestamp}.jpg"
                        val file = File(tempDir, filename)
                        file.outputStream().use { out ->
                            bitmap.compress(Bitmap.CompressFormat.JPEG, 95, out)
                        }
                        logger.info("StreamSession", "Photo saved (HEIC→JPEG)", mapOf("path" to file.absolutePath))

                        emitEvent("onPhotoCaptured", mapOf(
                            "filePath" to file.absolutePath,
                            "format" to "jpeg",
                            "timestamp" to timestamp,
                            "width" to bitmap.width,
                            "height" to bitmap.height
                        ))
                        return
                    }
                    logger.warn("StreamSession", "HEIC→JPEG conversion failed, saving as HEIC")
                }

                // Save as HEIC (default or conversion failed)
                val filename = "emwdat_photo_${timestamp}.heic"
                val file = File(tempDir, filename)
                file.writeBytes(bytes)
                logger.info("StreamSession", "Photo saved (HEIC)", mapOf("path" to file.absolutePath))

                // Try to get dimensions from EXIF
                var width = 0
                var height = 0
                try {
                    val exif = ExifInterface(ByteArrayInputStream(bytes))
                    width = exif.getAttributeInt(ExifInterface.TAG_IMAGE_WIDTH, 0)
                    height = exif.getAttributeInt(ExifInterface.TAG_IMAGE_LENGTH, 0)
                } catch (e: Exception) {
                    logger.warn("StreamSession", "Could not read HEIC EXIF", mapOf("error" to e.toString()))
                }

                val payload = mutableMapOf<String, Any>(
                    "filePath" to file.absolutePath,
                    "format" to "heic",
                    "timestamp" to timestamp
                )
                if (width > 0 && height > 0) {
                    payload["width"] = width
                    payload["height"] = height
                }
                emitEvent("onPhotoCaptured", payload)
            }
            else -> {
                logger.warn("StreamSession", "Unknown PhotoData type")
            }
        }
    }

    // MARK: - Mapping Helpers

    private fun mapStreamState(state: StreamSessionState): String = when (state) {
        StreamSessionState.STOPPED -> "stopped"
        StreamSessionState.STARTING -> "starting"
        StreamSessionState.STARTED -> "starting"
        StreamSessionState.STREAMING -> "streaming"
        StreamSessionState.STOPPING -> "stopping"
        StreamSessionState.CLOSED -> "stopped"
    }

    // MARK: - Cleanup

    private fun destroyStream(sessionId: String) {
        videoJobs[sessionId]?.cancel()
        videoJobs.remove(sessionId)
        stateJobs[sessionId]?.cancel()
        stateJobs.remove(sessionId)
        errorJobs[sessionId]?.cancel()
        errorJobs.remove(sessionId)
        streams.remove(sessionId)
        logger.debug("StreamSession", "Stream destroyed", mapOf("sessionId" to sessionId))
    }

    fun destroy() {
        for (sessionId in streams.keys.toList()) {
            destroyStream(sessionId)
        }
    }

    // MARK: - Event Emission

    private fun emitEvent(name: String, body: Map<String, Any>) {
        eventEmitter?.invoke(name, body)
    }
}
